README.md
Backbone dualStorage Adapter v1.4.2
===================================

A dualStorage adapter for Backbone. It's a drop-in replacement for Backbone.Sync() to handle saving to a localStorage database as a cache for the remote models.

Usage
-----

Include Backbone.dualStorage after having included Backbone.js:

    <script type="text/javascript" src="backbone.js"></script>
    <script type="text/javascript" src="backbone.dualstorage.js"></script>

Create your models and collections in the usual way.
Feel free to use Backbone as you usually would; this is a drop-in replacement.

Keep in mind that Backbone.dualStorage really loves your models. By default it will cache everything that passes through Backbone.sync. You can override this behaviour with the booleans ```remote``` or ```local``` on models and collections:

    SomeCollection = Backbone.Collection.extend({
        remote: true // never cached, dualStorage is bypassed entirely
        local: true  // always fetched and saved only locally, never saves on remote
        local: function() { return trueOrFalse; } // local and remote can also be dynamic
    });

You can also deactivate dualsync to some requests, when you want to sync with the server only later.

    SomeCollection.create({name: "someone"}, {remote: false});

Data synchronization
--------------------

When the client goes offline, dualStorage allows you to keep changing and destroying records. All changes will be sent when the client goes online again.

    // server online. Go!
    people.fetch();       // load people models and save them into localstorage

    // server offline!
    people.create({name: "Turing"});   // you still can create new people...
    people.models[0].save({age: 41});  // update existing ones...
    people.models[1].destroy();        // and destroy as well

    // collections track what is dirty and destroyed
    people.dirtyModels()               // => Array of dirty models
    people.destroyedModelIds()         // => Array of destroyed model ids

    // server online again!
    people.syncDirtyAndDestroyed();    // all changes are sent to the server and localStorage is updated

Keep in mind that if you try to fetch() a collection that has dirty data, only data currently in the localStorage will be loaded. collection.syncDirtyAndDestroyed() needs to be executed before trying to download new data from the server.

It is possible to tell whether the operation succeeded remotely or locally by examining `options.dirty` in the `success` callback:

	model.save({
		name: "Turing"
	}, {
		success: function(model, response, options) {
			if (options.dirty) {
				// saved locally
			} else {
				// saved remotely
			}
		}
	});

Offline state detection
-----------------------
dualStorage **always** treats an Ajax status code of `0` as an indication it is working offline. Additional status codes can be added by setting `offlineStatusCodes` to either an array:

    Backbone.DualStorage.offlineStatusCodes = [408];

or a function that accepts the `response` object and returns an array:

    Backbone.DualStorage.offlineStatusCodes = function(xhr) {
        var codes = [];

        if (...) {
            codes.push(xhr.status);
        }

        return codes;
    }


Data parsing
------------

Sometimes you may want to customize how data from the remote server is parsed before it's saved to localStorage.
Typically your model's `parse` method takes care of this.
Since dualStorage provides two layers of backend, we need a second parse method.
For example, if your remote API returns data in a way that the default `parse` method interprets the result as a single record,
use `parseBeforeLocalSave` to break up the data into an array of records like you would with [parse](http://backbonejs.org/#Model-parse).

* The model's `parse` method still parses data read from localStorage.
* The model's `parseBeforeLocalSave` method parses data read from the remote _before_ it is saved to localStorage on read.

Local data storage
------------------

dualStorage stores the local cache in localStorage.
Each collection's (or model's) `url` property is used as the storage namespace to separate different collections of data.
This can be overridden by defining a `storeName` property on your model or collection.
Defining storeName can be useful when your url is dynamic or when your models do not have the collection set but should be treated as part of that collection in the local cache.

Install
-------

Clone like usual or via npm `npm install nilbus/backbone.dualstorage --save`.

Contributing
------------

See CONTRIBUTING.md

Authors
-------

Thanks to [Edward Anderson](https://github.com/nilbus) for the test suite and continued maintenance.
Thanks to [Lucian Mihaila](https://github.com/lucian1900) for creating Backbone.dualStorage.
Thanks to [Jerome Gravel-Niquet](https://github.com/jeromegn) for Backbone.localStorage.


CHANGES.md
### 1.4.2 / 2017-07-30

* Merge #156: Pass options to model creation - supports integration in systems with custom backbone models (@paaddyy)

### 1.4.1 / 2016-04-18

* Fix #145: Prevent model's id from being nulled (and saved locally with that id) when an offlineStatusCode is returned; broken since 1.4.0

### 1.4.0 / 2015-02-16

* Fix #123: Expose original BackboneSync as Backbone.DualStorage.originalSync to allow custom online sync methods (Richard Tibbles)
* Fix #117: Don't generate temporary IDs that conflict with UUIDs (David Almilli)
* Fix #115: Don't attempt to delete models from the server that were only local, and call the callback in this case (Micha Reiser)
* Fix bug introduced by #94 in 1.3.1: Allow records to be created when offline and no store exists
* Merge #123: Expose original Backbone Sync to allow for customization of online sync method (Richard Tibbles)
* Merge #132, #133: Allow options to be passed to toJSON (Pavel Karoukin)
* Allow Backbone.DualStorage.offlineStatusCodes to include 200 OK as an offline status
* syncDirty, syncDestroyed, and syncDirtyAndDestroyed accept options (and therefore callbacks) which are passed to save and destroy
* Call the error callback when offline and fetching a model that was never cached

### 1.3.1 / 2014-15-30

* Make syncDestroyed compatible with backbone >= 1.1.1
* Fix #94: Call the error callback when offline if the collection has never been fetched (Dave Taylor)
* Fix #99: Restore compatibility with lodash; incompatible since 1.2.0 (Aleksandr Motsjonov)

### 1.3.0 / 2014-05-10

* Fix #104, #67: Instead of treating ajax errors as offline, use the error callback (Elad Efrat)
* Fix #106: Restore proper call to a model's parse method; broken since 1.0.2 (Eduardo Matos)
* Fix #105: Always set options.dirty in callbacks when offline (Elad Efrat)
* Fix #93: syncDirty works when models use a custom idAttribute (Ben Salinas)
* Fix #78: Do not clear local collection cache when a model is fetched
* Add Backbone.DualStorage.offlineStatusCodes for configuring what to consider as offline
* Add `make watch` for continual coffeescript compilation during development
* `make` compiles sourcemaps between coffeescript and javascript
* Prevent id duplication in the internal list of model ids for a collection when a model is fetched

### 1.2.0 / 2014-01-31

* Add dirty/destroyed querying via `Collection.dirtyModels` and `Collection.destroyedModelIds`
* Allow the default url-based storeName to be overriden with a `storeName` property on the model or collection
* Use `model.urlRoot` as the storeName, when available, before `model.url`.
  This fixes the issue described in #80 where models with the same `urlRoot`
  that are intended to be part of the same collection end up in different stores
  when the collection attribute is not set on the model.

  **Existing apps that rely on this incorrect behavior may break.**

  If your app expects models with the same `urlRoot` and differing `url`s to
  be in different stores locally, use the following workaround in your model:

        storeName: function() { return this.url() }

* Ensure models in the dirty list exist before saving.
  This mitigates concurrency issues noted in #62 until #35 is resolved, which should fix this problem.
* Guard against JSON.parse(null) for Android browsers
* Remove all usages of Model.clone() to play along with plugins (backbone-relational) that do not work with clone.
* Fix where fetching models/collections would not merge but overwrite locally stored attributes.
* Use the model idAttribute when saving models that were created offline.
  In this scenario, an update request (for an object with a temp ID) would be sent on save instead of a create request.

### 1.1.0 / 2013-11-10

* Add support for RequireJS / AMD

### 1.0.2 / 2013-11-10

* Fix support for models with a custom `idAttribute`
* Update locally cached attributes when the server response to a save updates attributes
* Add bower metadata
* Support non-numeric model ids
* Conform localSync to Backbone.sync protocol by returning attributes
* Fix support for models with `url` defined as a function

### 1.0.1 / 2013-03-27

* Add compatibility for Backbone 0.9.10
* Remove console.log calls for cleanup and IE support
* Add build Makefile for development
* Add test suite
* Add limited supoort for `fetch` add and merge options
* Support defining `parseBeforeLocalSave` on models for parsing server responses before passing them to localsync
* Support controlling behavior with `local`  and `remote` options

### 1.0.0 / 2013-01-27

* Forked Backbone.dualStorage from Backbone.localStorage


backbone.dualstorage.amd.js
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    define(['backbone'], factory);
  } else if (typeof require === 'function' && ((typeof module !== "undefined" && module !== null ? module.exports : void 0) != null)) {
    return module.exports = factory(require('backbone'));
  } else {
    factory(root.Backbone);
  }
})(this, function(Backbone) {
// Generated by CoffeeScript 1.9.3

/*
Backbone dualStorage Adapter v1.4.2

A simple module to replace `Backbone.sync` with *localStorage*-based
persistence. Models are given GUIDS, and saved into a JSON object. Simple
as that.
 */
var S4, backboneSync, callbackTranslator, dualsync, getStoreName, localsync, modelUpdatedWithResponse, onlineSync, parseRemoteResponse,
  indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

Backbone.DualStorage = {
  offlineStatusCodes: [408, 502]
};

Backbone.Model.prototype.hasTempId = function() {
  return _.isString(this.id) && this.id.length === 36 && this.id.indexOf('t') === 0;
};

getStoreName = function(collection, model) {
  model || (model = collection.model.prototype);
  return _.result(collection, 'storeName') || _.result(model, 'storeName') || _.result(collection, 'url') || _.result(model, 'urlRoot') || _.result(model, 'url');
};

Backbone.Collection.prototype.syncDirty = function(options) {
  var i, id, ids, len, ref, results, store;
  store = localStorage.getItem((getStoreName(this)) + "_dirty");
  ids = (store && store.split(',')) || [];
  results = [];
  for (i = 0, len = ids.length; i < len; i++) {
    id = ids[i];
    results.push((ref = this.get(id)) != null ? ref.save(null, options) : void 0);
  }
  return results;
};

Backbone.Collection.prototype.dirtyModels = function() {
  var id, ids, models, store;
  store = localStorage.getItem((getStoreName(this)) + "_dirty");
  ids = (store && store.split(',')) || [];
  models = (function() {
    var i, len, results;
    results = [];
    for (i = 0, len = ids.length; i < len; i++) {
      id = ids[i];
      results.push(this.get(id));
    }
    return results;
  }).call(this);
  return _.compact(models);
};

Backbone.Collection.prototype.syncDestroyed = function(options) {
  var i, id, ids, len, model, results, store;
  store = localStorage.getItem((getStoreName(this)) + "_destroyed");
  ids = (store && store.split(',')) || [];
  results = [];
  for (i = 0, len = ids.length; i < len; i++) {
    id = ids[i];
    model = new this.model;
    model.set(model.idAttribute, id);
    model.collection = this;
    results.push(model.destroy(options));
  }
  return results;
};

Backbone.Collection.prototype.destroyedModelIds = function() {
  var ids, store;
  store = localStorage.getItem((getStoreName(this)) + "_destroyed");
  return ids = (store && store.split(',')) || [];
};

Backbone.Collection.prototype.syncDirtyAndDestroyed = function(options) {
  this.syncDirty(options);
  return this.syncDestroyed(options);
};

S4 = function() {
  return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
};

window.Store = (function() {
  Store.prototype.sep = '';

  function Store(name) {
    this.name = name;
    this.records = this.recordsOn(this.name);
  }

  Store.prototype.generateId = function() {
    return 't' + S4().substring(1) + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4();
  };

  Store.prototype.save = function() {
    return localStorage.setItem(this.name, this.records.join(','));
  };

  Store.prototype.recordsOn = function(key) {
    var store;
    store = localStorage.getItem(key);
    return (store && store.split(',')) || [];
  };

  Store.prototype.dirty = function(model) {
    var dirtyRecords;
    dirtyRecords = this.recordsOn(this.name + '_dirty');
    if (!_.include(dirtyRecords, model.id.toString())) {
      dirtyRecords.push(model.id);
      localStorage.setItem(this.name + '_dirty', dirtyRecords.join(','));
    }
    return model;
  };

  Store.prototype.clean = function(model, from) {
    var dirtyRecords, store;
    store = this.name + "_" + from;
    dirtyRecords = this.recordsOn(store);
    if (_.include(dirtyRecords, model.id.toString())) {
      localStorage.setItem(store, _.without(dirtyRecords, model.id.toString()).join(','));
    }
    return model;
  };

  Store.prototype.destroyed = function(model) {
    var destroyedRecords;
    destroyedRecords = this.recordsOn(this.name + '_destroyed');
    if (!_.include(destroyedRecords, model.id.toString())) {
      destroyedRecords.push(model.id);
      localStorage.setItem(this.name + '_destroyed', destroyedRecords.join(','));
    }
    return model;
  };

  Store.prototype.create = function(model, options) {
    if (!_.isObject(model)) {
      return model;
    }
    if (!model.id) {
      model.set(model.idAttribute, this.generateId());
    }
    localStorage.setItem(this.name + this.sep + model.id, JSON.stringify(model.toJSON ? model.toJSON(options) : model));
    this.records.push(model.id.toString());
    this.save();
    return model;
  };

  Store.prototype.update = function(model, options) {
    localStorage.setItem(this.name + this.sep + model.id, JSON.stringify(model.toJSON ? model.toJSON(options) : model));
    if (!_.include(this.records, model.id.toString())) {
      this.records.push(model.id.toString());
    }
    this.save();
    return model;
  };

  Store.prototype.clear = function() {
    var i, id, len, ref;
    ref = this.records;
    for (i = 0, len = ref.length; i < len; i++) {
      id = ref[i];
      localStorage.removeItem(this.name + this.sep + id);
    }
    this.records = [];
    return this.save();
  };

  Store.prototype.hasDirtyOrDestroyed = function() {
    return !_.isEmpty(localStorage.getItem(this.name + '_dirty')) || !_.isEmpty(localStorage.getItem(this.name + '_destroyed'));
  };

  Store.prototype.find = function(model) {
    var modelAsJson;
    modelAsJson = localStorage.getItem(this.name + this.sep + model.id);
    if (modelAsJson === null) {
      return null;
    }
    return JSON.parse(modelAsJson);
  };

  Store.prototype.findAll = function() {
    var i, id, len, ref, results;
    ref = this.records;
    results = [];
    for (i = 0, len = ref.length; i < len; i++) {
      id = ref[i];
      results.push(JSON.parse(localStorage.getItem(this.name + this.sep + id)));
    }
    return results;
  };

  Store.prototype.destroy = function(model) {
    localStorage.removeItem(this.name + this.sep + model.id);
    this.records = _.reject(this.records, function(record_id) {
      return record_id === model.id.toString();
    });
    this.save();
    return model;
  };

  return Store;

})();

window.Store.exists = function(storeName) {
  return localStorage.getItem(storeName) !== null;
};

callbackTranslator = {
  needsTranslation: Backbone.VERSION === '0.9.10',
  forBackboneCaller: function(callback) {
    if (this.needsTranslation) {
      return function(model, resp, options) {
        return callback.call(null, resp);
      };
    } else {
      return callback;
    }
  },
  forDualstorageCaller: function(callback, model, options) {
    if (this.needsTranslation) {
      return function(resp) {
        return callback.call(null, model, resp, options);
      };
    } else {
      return callback;
    }
  }
};

localsync = function(method, model, options) {
  var isValidModel, preExisting, response, store;
  isValidModel = (method === 'clear') || (method === 'hasDirtyOrDestroyed');
  isValidModel || (isValidModel = model instanceof Backbone.Model);
  isValidModel || (isValidModel = model instanceof Backbone.Collection);
  if (!isValidModel) {
    throw new Error('model parameter is required to be a backbone model or collection.');
  }
  store = new Store(options.storeName);
  response = (function() {
    switch (method) {
      case 'read':
        if (model instanceof Backbone.Model) {
          return store.find(model);
        } else {
          return store.findAll();
        }
        break;
      case 'hasDirtyOrDestroyed':
        return store.hasDirtyOrDestroyed();
      case 'clear':
        return store.clear();
      case 'create':
        if (options.add && !options.merge && (preExisting = store.find(model))) {
          return preExisting;
        } else {
          model = store.create(model, options);
          if (options.dirty) {
            store.dirty(model);
          }
          return model;
        }
        break;
      case 'update':
        store.update(model, options);
        if (options.dirty) {
          return store.dirty(model);
        } else {
          return store.clean(model, 'dirty');
        }
        break;
      case 'delete':
        store.destroy(model);
        if (options.dirty && !model.hasTempId()) {
          return store.destroyed(model);
        } else {
          if (model.hasTempId()) {
            return store.clean(model, 'dirty');
          } else {
            return store.clean(model, 'destroyed');
          }
        }
    }
  })();
  if (response) {
    if (response.toJSON) {
      response = response.toJSON(options);
    }
    if (response.attributes) {
      response = response.attributes;
    }
  }
  if (!options.ignoreCallbacks) {
    if (response) {
      options.success(response);
    } else {
      options.error('Record not found');
    }
  }
  return response;
};

parseRemoteResponse = function(object, response) {
  if (!(object && object.parseBeforeLocalSave)) {
    return response;
  }
  if (_.isFunction(object.parseBeforeLocalSave)) {
    return object.parseBeforeLocalSave(response);
  }
};

modelUpdatedWithResponse = function(model, response) {
  var modelClone;
  modelClone = new Backbone.Model;
  modelClone.idAttribute = model.idAttribute;
  modelClone.set(model.attributes);
  modelClone.set(model.parse(response));
  return modelClone;
};

backboneSync = Backbone.DualStorage.originalSync = Backbone.sync;

onlineSync = function(method, model, options) {
  options.success = callbackTranslator.forBackboneCaller(options.success);
  options.error = callbackTranslator.forBackboneCaller(options.error);
  return Backbone.DualStorage.originalSync(method, model, options);
};

dualsync = function(method, model, options) {
  var error, hasOfflineStatusCode, local, relayErrorCallback, success, temporaryId, useOfflineStorage;
  options.storeName = getStoreName(model.collection, model);
  options.storeExists = Store.exists(options.storeName);
  options.success = callbackTranslator.forDualstorageCaller(options.success, model, options);
  options.error = callbackTranslator.forDualstorageCaller(options.error, model, options);
  if (_.result(model, 'remote') || _.result(model.collection, 'remote')) {
    return onlineSync(method, model, options);
  }
  local = _.result(model, 'local') || _.result(model.collection, 'local');
  options.dirty = options.remote === false && !local;
  if (options.remote === false || local) {
    return localsync(method, model, options);
  }
  options.ignoreCallbacks = true;
  success = options.success;
  error = options.error;
  useOfflineStorage = function() {
    options.dirty = true;
    options.ignoreCallbacks = false;
    options.success = success;
    options.error = error;
    return localsync(method, model, options);
  };
  hasOfflineStatusCode = function(xhr) {
    var offlineStatusCodes, ref;
    offlineStatusCodes = Backbone.DualStorage.offlineStatusCodes;
    if (_.isFunction(offlineStatusCodes)) {
      offlineStatusCodes = offlineStatusCodes(xhr);
    }
    return xhr.status === 0 || (ref = xhr.status, indexOf.call(offlineStatusCodes, ref) >= 0);
  };
  relayErrorCallback = function(xhr) {
    var online;
    online = !hasOfflineStatusCode(xhr);
    if (online || method === 'read' && !options.storeExists) {
      return error(xhr);
    } else {
      return useOfflineStorage();
    }
  };
  switch (method) {
    case 'read':
      if (localsync('hasDirtyOrDestroyed', model, options)) {
        return useOfflineStorage();
      } else {
        options.success = function(resp, _status, _xhr) {
          var collection, i, idAttribute, len, modelAttributes, responseModel;
          if (hasOfflineStatusCode(options.xhr)) {
            return useOfflineStorage();
          }
          resp = parseRemoteResponse(model, resp);
          if (model instanceof Backbone.Collection) {
            collection = model;
            idAttribute = collection.model.prototype.idAttribute;
            if (!options.add) {
              localsync('clear', collection, options);
            }
            for (i = 0, len = resp.length; i < len; i++) {
              modelAttributes = resp[i];
              model = collection.get(modelAttributes[idAttribute]);
              if (model) {
                responseModel = modelUpdatedWithResponse(model, modelAttributes);
              } else {
                responseModel = new collection.model(modelAttributes, options);
              }
              localsync('update', responseModel, options);
            }
          } else {
            responseModel = modelUpdatedWithResponse(model, resp);
            localsync('update', responseModel, options);
          }
          return success(resp, _status, _xhr);
        };
        options.error = function(xhr) {
          return relayErrorCallback(xhr);
        };
        return options.xhr = onlineSync(method, model, options);
      }
      break;
    case 'create':
      options.success = function(resp, _status, _xhr) {
        var updatedModel;
        if (hasOfflineStatusCode(options.xhr)) {
          return useOfflineStorage();
        }
        updatedModel = modelUpdatedWithResponse(model, resp);
        localsync(method, updatedModel, options);
        return success(resp, _status, _xhr);
      };
      options.error = function(xhr) {
        return relayErrorCallback(xhr);
      };
      return options.xhr = onlineSync(method, model, options);
    case 'update':
      if (model.hasTempId()) {
        temporaryId = model.id;
        options.success = function(resp, _status, _xhr) {
          var updatedModel;
          model.set(model.idAttribute, temporaryId, {
            silent: true
          });
          if (hasOfflineStatusCode(options.xhr)) {
            return useOfflineStorage();
          }
          updatedModel = modelUpdatedWithResponse(model, resp);
          localsync('delete', model, options);
          localsync('create', updatedModel, options);
          return success(resp, _status, _xhr);
        };
        options.error = function(xhr) {
          model.set(model.idAttribute, temporaryId, {
            silent: true
          });
          return relayErrorCallback(xhr);
        };
        model.set(model.idAttribute, null, {
          silent: true
        });
        return options.xhr = onlineSync('create', model, options);
      } else {
        options.success = function(resp, _status, _xhr) {
          var updatedModel;
          if (hasOfflineStatusCode(options.xhr)) {
            return useOfflineStorage();
          }
          updatedModel = modelUpdatedWithResponse(model, resp);
          localsync(method, updatedModel, options);
          return success(resp, _status, _xhr);
        };
        options.error = function(xhr) {
          return relayErrorCallback(xhr);
        };
        return options.xhr = onlineSync(method, model, options);
      }
      break;
    case 'delete':
      if (model.hasTempId()) {
        options.ignoreCallbacks = false;
        return localsync(method, model, options);
      } else {
        options.success = function(resp, _status, _xhr) {
          if (hasOfflineStatusCode(options.xhr)) {
            return useOfflineStorage();
          }
          localsync(method, model, options);
          return success(resp, _status, _xhr);
        };
        options.error = function(xhr) {
          return relayErrorCallback(xhr);
        };
        return options.xhr = onlineSync(method, model, options);
      }
  }
};

Backbone.sync = dualsync;

//# sourceMappingURL=backbone.dualstorage.js.map
});